### ISSUE TITLE
feature: add assets dynamically

### ISSUE THREAD

**Issue by lispc** [Fluidex/dingir-exchange#8]
Currently the matching server loads available trading pairs ( like ETH/USDT ) from the config.yaml file when it starts.    

We need a new API to add a new trading pair on the fly.  

My naive design ( depends <https://github.com/Fluidex/dingir-exchange/issues/77> ) 

1. add a new rpc endpoint in the `restapi` process,  which first add a new trading pair to the db, then call 'ReloadMarkets' grpc of matchengine
2. in the `ReloadMarkets` grpc endpoint of `matchengine` server, it "select * from market where create_time > ${last_market_create_time}". Then it initialize new markets, and **set last_market_create_time to create_time of the newest market**. By using "last_market_create_time", the matchengine avoid reloading all the markets everytime. It only need to reload new markets.

**Comment by HAOYUatHZ** [comment]
> add a new rpc endpoint, which can add a new trading pair to the db.

I am not sure whether it's a good idea. Maybe we can use https://github.com/ctripcorp/apollo?

**Comment by HAOYUatHZ** [comment]
At the moment maybe using redis for hot configuring `trading_pairs` is already sufficient.

**Comment by lispc** [comment]
I do not know much of the cloud-native stuff. Using redis seems sufficient and simple

**Comment by HAOYUatHZ** [comment]
> ~I do not know much of the cloud-native stuff. Using redis seems sufficient and simple~
> 
> The API is used for permissionless token listing. It seems a 'reload' API is needed even using redis.

If we simply put it in redis, we can have:

+ an apollo running for hot configuration
+ an apollo client to fecth configuration from apollo and update it into redis 
+ dingir-exchange server fetches config from redis whenever needed.

These will also support other and flexible format config.

`/trading_pairs` looks a frequently used API, we then need to put it into memory for saving redis workload and saving overhead.
**Either 1) your `\reload` API or 2) letting dingir-exchange server fetches from redis every 5 min would work.**

The main concern is to design an `add_token` mechanism so that people can add token without modifying existing tokens. Then it seems that we still need a db (PG here because we already have a PG) for looking up.  

So, in conclusion, what I am thinking about for "trading_pairs" is:
+ add a new table named `markets` or `trading_pairs`.
+ provide a client/rpc_endpoint for users to add token. This doesn't need authentication if aiming at permissionless token listing.
    + we will also need to limit the access rate.
+ server fetch "trading_pairs" from db every 5 minutes.

We can switch to apollo (there's no apollo rust client, so we may still need a redis and add authentication to the redis) in the future if we need more flexible configurations for more complex scenario.

**Comment by lispc** [comment]
I agree with this design.

`This doesn't need authentication if aiming at permissionless token listing.` : `add_token` may not be a public rpc endpoint. One of our process will monitor eth blocks/txs/events, when a user finished an online tx 'addToken' and an event was emitted, the monitor process calls the the rpc to add the token into redis config. Then within at most 5 minutes, the matching engine reloads , and everything works.

**Comment by HAOYUatHZ** [comment]
> add_token may not be a public rpc endpoint. One of our process will monitor eth blocks/txs/events, when a user finished an online tx 'addToken' and an event was emitted, the monitor process calls the the rpc to add the token into redis config. Then within at most 5 minutes, the matching engine reloads , and everything works.

Looks great!

**Comment by HAOYUatHZ** [comment]
>  the monitor process calls the the rpc to add the token into redis config. Then within at most 5 minutes, the matching engine reloads , and everything works.

For now  we don't even need redis. We can just let the monitor process, by or not by RPC, check in PG and then insert into PG.

**Comment by HAOYUatHZ** [comment]
We still need PG for persisting trading_pairs

**Comment by noel2004** [comment]
> Currently the matching server loads available trading pairs ( like ETH/USDT ) from the config.yaml file when it starts.
> 
> We need a new API to add a new trading pair on the fly.
> 
> My naive design ( depends #77 )
> 
> 1. add a new rpc endpoint in the `restapi` process,  which first add a new trading pair to the db, then call 'ReloadMarkets' grpc of matchengine
> 2. in the `ReloadMarkets` grpc endpoint of `matchengine` server, it "select * from market where create_time > ${last_market_create_time}". Then it initialize new markets, and **set last_market_create_time to create_time of the newest market**. By using "last_market_create_time", the matchengine avoid reloading all the markets everytime. It only need to reload new markets.

A rpc enpoint ("add_token"), at least for the debug purpose, seems useful. So first I will work base on this description.

**Comment by HAOYUatHZ** [comment]
closed via #85
